utils/defaulter.go

```diff
@@ -126,3 +155,143 @@ func (d *Defaulter) MustSet(arg interface{}) {
 		panic(err)
 	}
 }
+
+func (d *Defaulter) getEntitySchema(entityType string) (map[string]interface{}, error) {
+	var schema map[string]interface{}
+	schema, err := d.client.Schemas.Get(d.ctx, entityType)
+	if err != nil {
+		return schema, err
+	}
+	return schema, nil
+}
+
+func (d *Defaulter) addEntityDefaults(entityType string, entity interface{}) error {
+	schema, err := d.getEntitySchema(entityType)
+	if err != nil {
+		return fmt.Errorf("retrieve schema for %v from Kong: %v", entityType, err)
+	}
+	return kong.FillEntityDefaults(entity, schema)
+}
+
+func getKongDefaulterWithClient(ctx context.Context, client *kong.Client,
+	kongDefaults interface{}) (*Defaulter, error) {
+	// fills defaults from input
+	d, err := GetKongDefaulter(kongDefaults, false)
+	if err != nil {
+		return nil, err
+	}
+	d.ctx = ctx
+	d.client = client
+
+	// fills defaults from Kong API
+	if err := d.addEntityDefaults("services", d.service); err != nil {
+		return nil, fmt.Errorf("get defaults for services: %v", err)
+	}
+	if err := d.Register(d.service); err != nil {
+		return nil, fmt.Errorf("registering service with defaulter: %w", err)
+	}
+
+	if err := d.addEntityDefaults("routes", d.route); err != nil {
+		return nil, fmt.Errorf("get defaults for routes: %v", err)
+	}
+	if err := d.Register(d.route); err != nil {
+		return nil, fmt.Errorf("registering route with defaulter: %w", err)
+	}
+
+	if err := d.addEntityDefaults("upstreams", d.upstream); err != nil {
+		return nil, fmt.Errorf("get defaults for upstreams: %v", err)
+	}
+	if err := d.Register(d.upstream); err != nil {
+		return nil, fmt.Errorf("registering upstream with defaulter: %w", err)
+	}
+
+	if err := d.addEntityDefaults("targets", d.target); err != nil {
+		return nil, fmt.Errorf("get defaults for targets: %v", err)
+	}
+	if err := d.Register(d.target); err != nil {
+		return nil, fmt.Errorf("registering target with defaulter: %w", err)
+	}
+	return d, nil
+}
+
+// GetDefaulter returns a Defaulter object to be used to set defaults
+// on Kong entities. The order of precedence is as follow, from higher to lower:
+//
+// 1. values set in the state file
+// 2. values set in the {_info: defaults:} object in the state file
+// 3. schema defaults coming from Admin API (excluded Konnect)
+// 4. hardcoded defaults under utils/constants.go (Konnect-only)
+func GetDefaulter(
+	ctx context.Context, client *kong.Client, kongDefaults interface{}, isKonnect bool,
```

Comment:
Take in all args as options, except ctx.

Reply:
done